cdc/sinkv2/eventsink/factory/factory.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package factory
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink"
+	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq"
+	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/dmlproducer"
+	"github.com/pingcap/tiflow/cdc/sinkv2/tablesink"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/kafka"
+)
+
+// sinkType is the type of sink.
+type sinkType int
+
+const (
+	mqSink sinkType = iota + 1
+	txnSink
+)
+
+// SinkFactory is the factory of sink.
+// It is responsible for creating sink and closing it.
+// Because there is no way to convert the eventsink.EventSink[*model.RowChangedEvent]
+// to eventsink.EventSink[eventsink.TableEvent].
+// So we have to use this factory to create and store the sink.
+type SinkFactory struct {
```

Comment:
IMO, it's not a clear abstract because we will `switch-case` everywhere. It's better to define the `SinkFactory` interface and implement the `MQSinkFactory` and `TxnSinkFactory` to make our code clean.

BTW, We just need to add a new implementation instead of change all the `switch-case` if we add more implementation for the interface `SinkFactory` interface.

Reply:
I think it's difficult to achieve it. Becase SinkFactory has to hold the real sink instead of `eventsink.EventSink[eventsink.TableSink]`. This is also the reason for this hacky PR.